File: examples/order/node/src/orders/orders.service.ts
import { Injectable, NotFoundException, Logger } from "@nestjs/common";
import { CreateOrderDto } from "./dto/create-order.dto";
import { UpdateOrderDto, OrderStatus } from "./dto/update-order.dto";
import { Order } from "./entities/order.entity";
import { EventsService } from "../events/events.service";
import { v4 as uuidv4 } from "uuid";

@Injectable()
export class OrdersService {
  private readonly logger = new Logger(OrdersService.name);
  private orders: Map<string, Order> = new Map();

  constructor(private eventsService: EventsService) {}

  async create(createOrderDto: CreateOrderDto): Promise<Order> {
    const orderId = uuidv4();
    const now = new Date();

    const order: Order = {
      id: orderId,
      status: OrderStatus.PENDING,
      customerId: createOrderDto.customerId,
      items: createOrderDto.items,
      createdAt: now,
      updatedAt: now,
    };

    this.orders.set(orderId, order);

    this.logger.log(
      `Order created: ${orderId} for customer: ${createOrderDto.customerId}`,
      "OrdersService"
    );

    // Publish order created event
    await this.eventsService.publishOrderCreated(order);

    return order;
  }

  async findAll(
    limit: number = 20,
    offset: number = 0
  ): Promise<{ data: Order[]; total: number; limit: number; offset: number }> {
    const allOrders = Array.from(this.orders.values());
    const data = allOrders.slice(offset, offset + limit);

    return {
      data,
      total: allOrders.length,
      limit,
      offset,
    };
  }

  async findOne(id: string): Promise<Order> {
    const order = this.orders.get(id);

    if (!order) {
      throw new NotFoundException(`Order with ID ${id} not found`);
    }

    return order;
  }

  async update(id: string, updateOrderDto: UpdateOrderDto): Promise<Order> {
    const order = await this.findOne(id);
    const previousStatus = order.status;

    order.status = updateOrderDto.status;
    order.updatedAt = new Date();

    this.orders.set(id, order);

    this.logger.log(
      `Order ${id} status updated from ${previousStatus} to ${updateOrderDto.status}`,
      "OrdersService"
    );

    // Publish order updated event
    await this.eventsService.publishOrderUpdated(order, previousStatus);

    return order;
  }

  // For testing purposes - this would normally be handled by a database
  async clear(): Promise<void> {
    this.orders.clear();
  }
}

File: examples/order/dotnet/Orders/OrdersService.cs
using OrdersService.Events;
using Serilog;

namespace OrdersService.Orders;

public interface IOrdersService
{
    Task<Order> CreateOrderAsync(OrderCreate orderCreate);
    Task<Order?> GetOrderAsync(string id);
    Task<Order?> UpdateOrderAsync(string id, OrderUpdate orderUpdate);
    Task<List<Order>> GetOrdersAsync(int limit = 10, int offset = 0);
}

public class OrdersService : IOrdersService
{
    private readonly Dictionary<string, Order> _orders = new();
    private readonly IEventService _eventService;
    private readonly ILogger<OrdersService> _logger;

    public OrdersService(IEventService eventService, ILogger<OrdersService> logger)
    {
        _eventService = eventService;
        _logger = logger;
    }

    public async Task<Order> CreateOrderAsync(OrderCreate orderCreate)
    {
        var order = new Order(
            Id: Guid.NewGuid().ToString(),
            Status: OrderStatus.Pending,
            CustomerId: orderCreate.CustomerId,
            Items: orderCreate.Items,
            CreatedAt: DateTime.UtcNow,
            UpdatedAt: DateTime.UtcNow
        );

        _orders[order.Id] = order;

        _logger.LogInformation("Order created: {OrderId} for customer {CustomerId}", 
            order.Id, order.CustomerId);

        // Publish order.created event
        await _eventService.PublishAsync("order.created", new
        {
            id = order.Id,
            customerId = order.CustomerId,
            items = order.Items,
            status = order.Status.ToString().ToLower()
        });

        return order;
    }

    public Task<Order?> GetOrderAsync(string id)
    {
        _orders.TryGetValue(id, out var order);
        return Task.FromResult(order);
    }

    public async Task<Order?> UpdateOrderAsync(string id, OrderUpdate orderUpdate)
    {
        if (!_orders.TryGetValue(id, out var existingOrder))
        {
            return null;
        }

        var updatedOrder = existingOrder with 
        { 
            Status = orderUpdate.Status,
            UpdatedAt = DateTime.UtcNow
        };

        _orders[id] = updatedOrder;

        _logger.LogInformation("Order updated: {OrderId} status changed to {Status}", 
            id, orderUpdate.Status);

        // Publish order.updated event
        await _eventService.PublishAsync("order.updated", new
        {
            id = updatedOrder.Id,
            status = updatedOrder.Status.ToString().ToLower()
        });

        return updatedOrder;
    }

    public Task<List<Order>> GetOrdersAsync(int limit = 10, int offset = 0)
    {
        var orders = _orders.Values
            .OrderByDescending(o => o.CreatedAt)
            .Skip(offset)
            .Take(limit)
            .ToList();

        return Task.FromResult(orders);
    }
}
